src/System.Data.OleDb/src/RowBinding.cs

```diff
@@ -0,0 +1,554 @@
+                }
+            }
+            finally {
+                if (mustRelease) {
+                    DangerousRelease();
+                }
+            }
+        }
+
+        internal void CloseFromConnection() {
+            _iaccessor = null;
+            _accessorHandle = ODB.DB_INVALID_HACCESSOR;
+        }
+
+        internal new void Dispose() {
+            ResetValues();
+
+            UnsafeNativeMethods.IAccessor iaccessor = _iaccessor;
+            IntPtr accessorHandle = _accessorHandle;
+
+            _iaccessor = null;
+            _accessorHandle = ODB.DB_INVALID_HACCESSOR;
+
+            if ((ODB.DB_INVALID_HACCESSOR != accessorHandle) && (null != iaccessor)) {
+                OleDbHResult hr;
+                int refcount;
+                hr = iaccessor.ReleaseAccessor(accessorHandle, out refcount);
+                if (hr < 0) { // ignore any error msgs
+                    SafeNativeMethods.Wrapper.ClearErrorInfo();
+                }
+            }
+
+            base.Dispose();
+        }
+
+        internal void ResetValues() {
+            if (_needToReset && _haveData) {
+                lock(this) { // prevent Dispose/ResetValues race condition
+
+                    bool mustRelease = false;
+                    RuntimeHelpers.PrepareConstrainedRegions();
+                    try {
+                        DangerousAddRef(ref mustRelease);
+
+                        ResetValues(DangerousGetHandle(), _iaccessor);
+                    }
+                    finally {
+                        if (mustRelease) {
+                            DangerousRelease();
+                        }
+                    }
+                }
+            }
+            else {
+                _haveData = false;
+            }
+#if DEBUG
+            // verify types that need reseting are not forgotton, since the code
+            // that sets this up is in dbbinding.cs, MaxLen { set; }
+            if (!_needToReset) {
+                Debug.Assert(0 <= _bindingCount && (_bindingCount * ODB.SizeOf_tagDBBINDING) < Length, "bad _bindingCount");
+                for (int i = 0; i < _bindingCount; ++i) {
+                    short wtype = ReadInt16((i * ODB.SizeOf_tagDBBINDING) + ODB.OffsetOf_tagDBBINDING_wType);
+                    switch(wtype) {
+                    case (NativeDBType.BYREF | NativeDBType.BYTES):
+                    case (NativeDBType.BYREF | NativeDBType.WSTR):
+                    case NativeDBType.PROPVARIANT:
+                    case NativeDBType.VARIANT:
+                    case NativeDBType.BSTR:
+                    case NativeDBType.HCHAPTER:
+                        Debug.Assert(false, "expected _needToReset");
+                        break;
+                    }
+                }
+            }
+#endif
+        }
+
+        // requires ReliabilityContract to be called by ReleaseHandle
+        [ReliabilityContract(Consistency.WillNotCorruptState, Cer.Success)]
```

Comment:
ReliabilityContract attributes can all be removed. I thought I did this clearly I missed some.